### PR TITLE
fix (btc): Improvement btc utxo parser

### PR DESCRIPTION
**Pull request by vrockz747**
PRF-7224: https://app.clickup.com/t/9002019994/PRF-7224
PRF-7200: https://app.clickup.com/t/9002019994/PRF-7200
PRF-7212: https://app.clickup.com/t/9002019994/PRF-7212
PRF-7229: https://app.clickup.com/t/9002019994/PRF-7229

Sample sdk script for testing:
```
import DeviceConnection, {
  updateLogger as updateLoggerHid,
} from '@cypherock/sdk-hw-hid';
import DeviceConnectionSerialport from '@cypherock/sdk-hw-serialport';
import { ManagerApp, updateLogger } from '@cypherock/sdk-app-manager';
import { updateLogger as updateLoggerCore } from '@cypherock/sdk-core';
import { IDeviceConnection } from '@cypherock/sdk-interfaces';
import * as bitcoinJsLib from 'bitcoinjs-lib';
import { setBitcoinJSLib, BtcApp, ISignTxnInputData } from '@cypherock/sdk-app-btc';
import { createServiceLogger } from './logger';

const run = async () => {
  updateLogger(createServiceLogger);
  updateLoggerCore(createServiceLogger);
  updateLoggerHid(createServiceLogger);

  setBitcoinJSLib(bitcoinJsLib);

  let connection: IDeviceConnection;

  try {
    connection = await DeviceConnection.create();
  } catch (error) {
    connection = await DeviceConnectionSerialport.create();
  }

  const managerApp = await ManagerApp.create(connection);
  const deviceInfo = await managerApp.getDeviceInfo();
  console.log(deviceInfo);

  const walletId = (await managerApp.getWallets());
  console.log('Printing all wallets in the device.');
  console.log(walletId.walletList);

  const txnvar = {
    inputs: [
      {
        prevIndex: 0,
        prevTxnId:
          'e1483376a1cd64027f386c2e28650536a13c813e955469f1f486dafb4f4b861d',
        prevTxn:
          '0200000002e5937831604534dfe013f9a6a983ae6d6bd853f42f235ce70ca5a7fe81643738000000006b483045022100ced63a4d44463475dacc752dc0a6f67b892eb4894ec99117c9fb04ac03b05bae022036c9985920b9d838858082223eaa480172d00a78bc556a0a8550f6f64cb613c60121020e7f64d7dad196fd8fe07d667b50aa2be77c79dd841136d0f0659be3aa8e5ba6ffffffff1ef0dba72077858b5388513263a62678a0bff7bc45dc04164dcafac673d9fdbd000000006b483045022100bb32a2ac921737e958da4e44ba0dcd60e85b5583dab6cbbeb2efc81ca0bf3b410220464415e8a18ee2612fccb034a977c80bc9575adb358aa982b25e3eaed42927bb012102c8a84d4370f37d079896b2f296947ecb9d3812ba074cd38d79f2dc1ef137781affffffff01601800000000000016001429f267742a8cea09dce9853d7c4c4653909f345a00000000',
        value: '6240',
        address: 'bc1q98exwap23n4qnh8fs57hcnzx2wgf7dz6up764t',
        changeIndex: 0, // dummy value
        addressIndex: 10, // dummy value
        sequence: 0xffffffff,
      },
    ],
    outputs: [
      {
        value: '3168',
        address: 'bc1qe5pjgzk9ml6r6xsze8zqz73taehms4e59ttpdf',
        isChange: false,
      },
    ],
  }
  await (await BtcApp.create(connection)).signTxn({ derivationPath: [0x80000000 + 44, 0x80000000, 0x80000000], txn: txnvar, walletId: walletId.walletList[0].id });
  connection.destroy();
};

run();
```

### CODE REVIEW

**Review by ParnikaGupta9 — CHANGES_REQUESTED** [review]

File: tests/apps/btc_app/btc_txn_helpers_tests.c [hunk 1]
```diff
@@ -201,7 +252,33 @@ TEST(btc_txn_helper_test, btc_txn_helper_verify_input_p2pkh) {
...
+    for (int i = 0; i < (chvar); i++) {
+      txn_chunk[i] = raw_txn[i + (CHUNK_SIZE * index)];
+    }
+
+    status = btc_verify_input(txn_chunk, index, &verify_input_data, &input);
+    if (status != 4) {
+      break;
+    }
+    index++;
+  }
```
[the remark this answers is not in the document]
Comment (vrockz747) [reply tests/apps/btc_app/btc_txn_helpers_tests.c]: Fixed it in commit 2af3ba3553120bb60c28c1f4d5e094e363a3c686

File: apps/btc_family/btc_txn.c [hunk 2]
```diff
@@ -414,6 +405,86 @@ static bool fetch_valid_input(btc_query_t *query) {
...
+    // Free txin after use
+    free(txin);
+    if (total_size != size) {
+      btc_send_error(ERROR_COMMON_ERROR_CORRUPT_DATA_TAG,
+                     ERROR_DATA_FLOW_INVALID_DATA);
+      return false;
+    }
+
+    if ((SCRIPT_TYPE_P2PKH != type && SCRIPT_TYPE_P2WPKH != type) ||
+        0 != status) {
```
[the remark this answers is not in the document]
Comment (vrockz747) [reply apps/btc_family/btc_txn.c:482]: could you explain more, didn't get what you are referring to

File: apps/btc_family/btc_txn_helpers.c [hunk 3]
```diff
@@ -328,65 +328,245 @@ STATIC bool calculate_p2wpkh_digest(const btc_txn_context_t *context,
...
-    offset += 36;
-    offset += (raw_txn[offset] + 1 + 4);
+  if (chunk_index == 0) {
+    // ignore network version (4-bytes), skip marker & flag (in segwit)
+    offset += (raw_txn_chunk[4] == 0 ? 6 : 4);
+    if (6 == offset) {
+      verify_input_data->isSegwit = true;
+    }
+    // remember the number of inputs in the raw_txn
+    // TODO: UPDATE VAR-INT DECODE
```
[the remark this answers is not in the document]
Comment (vrockz747) [reply apps/btc_family/btc_txn_helpers.c:414]: current method assumes the `input-count` variable to be of single byte, which is of type `varint` and can varry upto 9 bytes as [specified.](https://en.bitcoin.it/wiki/Transaction).
This limitation of varint deocde is continued from the previous method and can be improved in future if required.

File: tests/apps/btc_app/btc_txn_helpers_tests.c [hunk 4]
```diff
@@ -100,32 +133,37 @@ TEST(btc_txn_helper_test, btc_txn_helper_verify_input_p2pk) {
...
+      "2d028510c68a024830450221008e2f802e44bdff18531a05ddc5b3788e84ca96b8136772"
+      "96dae224687d25df320220518d296d04cb9b40c338bf83afd07e73e810b90213c8751377"
+      "40dbebcba7af3b0121024aece25573da4ed274ee891341490fe26bd021b63e54e8581fc3"
+      "ca10085a8fa000000000",
+      452,
       raw_txn);
   // only fill necessary values
   btc_sign_txn_input_t input = {.prev_output_index = 0,
-                                .value = 5000000000,
+                                .value = 5000000000,    // Wrong value
```
[the remark this answers is not in the document]
Comment (vrockz747) [reply tests/apps/btc_app/btc_txn_helpers_tests.c:147]: yes the wrong value comment declares that the check is intended and expected to fail with return code of 3

File: tests/apps/btc_app/btc_txn_helpers_tests.c [hunk 5]
```diff
@@ -100,32 +133,37 @@ TEST(btc_txn_helper_test, btc_txn_helper_verify_input_p2pk) {
...
+      "01000000000101fee6b227e3e24bf2e5a4ea1ca6eba722c5eab233832c9c84b4893339ad"
+      "9294413200000000ffffffff02dcb93400000000001976a9141499db6b8f427c624a644b"
+      "f9442b9e3ec37f211d88aca9790100000000001600143026421e3930e0243b626d1445fb"
+      "2d028510c68a024830450221008e2f802e44bdff18531a05ddc5b3788e84ca96b8136772"
+      "96dae224687d25df320220518d296d04cb9b40c338bf83afd07e73e810b90213c8751377"
+      "40dbebcba7af3b0121024aece25573da4ed274ee891341490fe26bd021b63e54e8581fc3"
+      "ca10085a8fa000000000",
+      452,
       raw_txn);
   // only fill necessary values
```
[the remark this answers is not in the document]
Comment (vrockz747) [reply tests/apps/btc_app/btc_txn_helpers_tests.c:145]: replaced a test case of legacy type to segwit type to test compatibility with segwit txns.
should add more comments to this, or add this entirely as a new test case?

[the remark this answers is not in the document]
Comment (ParnikaGupta9) [reply tests/apps/btc_app/btc_txn_helpers_tests.c:145]: yes

**Review by ParnikaGupta9 — APPROVED** [review]

**Pull request merged**